File: in_Java/0130 Surrounded Regions.java
class Solution {
    // start from the 'O' on boarder and go inward
    public void solve(char[][] board) {
        for(int i = 0; i < board.length; i++){
            for(int j = 0; j < board[0].length; j++){
                if(i == 0 || j == 0 || i == board.length-1 || j == board[0].length-1){
                    if(board[i][j] == 'O') dfs(board,i,j);
                }
            }
        }
        for(int i = 0; i < board.length; i++){
            for(int j = 0; j < board[0].length; j++){
                if(board[i][j] == 'O') board[i][j] = 'X';
                if(board[i][j] == '*') board[i][j] = 'O';
            }
        }
    }
    private void dfs(char[][] g, int i, int j){
        if(i < 0 || j < 0 || i >= g.length || j >= g[0].length || g[i][j] == 'X' || g[i][j] == '*') return;
        g[i][j] = '*';
        dfs(g,i-1,j);
        dfs(g,i+1,j);
        dfs(g,i,j-1);
        dfs(g,i,j+1);        
    }
}

File: in_Python/0130 Surrounded Regions.py
class Solution(object):
    def solve(self, board):
        """
        :type board: List[List[str]]
        :rtype: None Do not return anything, modify board in-place instead.
        """
        if board == None:
            return board
        if len(board) == 0 or len(board[0]) == 0:
            return board
        row = len(board)
        col = len(board[0])
        directions = [[0,1],[0,-1],[1,0],[-1,0]]
        zeros_on_border = []
        for i in [0, row-1]:
            for j in range(col):
                if board[i][j] == 'O':
                    zeros_on_border.append([i,j])
        for j in [0, col-1]:
            for i in range(row):
                if board[i][j] == 'O' and [i,j] not in zeros_on_border:
                    zeros_on_border.append([i,j])
        while len(zeros_on_border) != 0:
            h = zeros_on_border.pop(0)
            thisi = h[0]
            thisj = h[1]
            board[thisi][thisj] = 'S'
            for d in directions:
                nexti = thisi + d[0]
                nextj = thisj + d[1]
                if 0 < nexti < row-1 and 0 < nextj < col-1:
                    if board[nexti][nextj] == 'O':
                        zeros_on_border.append([nexti,nextj])
                        board[nexti][nextj] = 'S'
        for i in range(row):
            for j in range(col):
                if board[i][j] == 'O':
                    board[i][j] = 'X'
                if board[i][j] == 'S':
                    board[i][j] = 'O'
        
            
        